Sources/NIO/Socket.swift

```diff
@@ -170,6 +170,59 @@ typealias IOVector = iovec
         }
     }
 
+    func recvmsg(pointer: UnsafeMutableRawBufferPointer, storage: inout sockaddr_storage, storageLen: inout socklen_t,
+                 controlBytes: inout Slice<UnsafeMutableRawBufferPointer>,
+                 controlMessageReceiver: ((Int32, Int32, UnsafeBufferPointer<UInt8>?) -> ())?) throws -> IOResult<Int> {
+        var vec = iovec(iov_base: pointer.baseAddress, iov_len: pointer.count)
+        let localControlBytePointer = UnsafeMutableRawBufferPointer(rebasing: controlBytes)
+
+        return try withUnsafeMutablePointer(to: &vec) { vecPtr in
+            return try storage.withMutableSockAddr { (sockaddrPtr, _) in
+                var messageHeader = msghdr(msg_name: sockaddrPtr,
+                                           msg_namelen: storageLen,
+                                           msg_iov: vecPtr,
+                                           msg_iovlen: 1,
+                                           msg_control: localControlBytePointer.baseAddress,
+                                           msg_controllen: .init(localControlBytePointer.count),
+                                           msg_flags: 0)
+                defer {
+                    // We need to write back the length of the message and the control bytes.
+                    storageLen = messageHeader.msg_namelen
+                    controlBytes = controlBytes.prefix(.init(messageHeader.msg_controllen))
+                    
+                }
+
+                let result = try withUnsafeMutablePointer(to: &messageHeader) { messageHeader in
+                    return try withUnsafeHandle { fd in
+                        return try Posix.recvmsg(descriptor: fd, msgHdr: messageHeader, flags: 0)
+                    }
+                }
+                
+                // Only look at the control bytes if all is good.
+                if case .processed = result {
+                    controlMessageReceiver.map { foreachControlMessage(in: messageHeader, $0)}
+                }
+                
+                return result
+            }
+        }
+    }
+    
+    private func foreachControlMessage(in messageHeader: msghdr,
```

Comment:
It would be much nicer to design this to have a Swift wrapper type around the control message data. The properties of this type would be:

1. It would be a `Collection`. The `Index` type of this collection would be a pointer to the individual control message, and it would be able to implement `index(after:)` by way of `CMSG_NXTHDR`.
2. The element types would be a type that provides the level, type, and pointer to the data. This type should itself be named and have names for these two fields (mystery tuples are gross).

This then allows us to reimplement `foreachControlMessage` with simple straightforward iteration.